File: features/miso/loop/cards/browse/map/basemap/boundaries/region/region.js
const feature_Region = {
  // the region, drawn twice: a second ground under a mask shaped like the
  // chosen polygon, and a page of pills to choose it with. Everything about
  // WHICH region comes from /boundaries' geojson and from the `#misoRegion`
  // marker the Rust half writes — nothing here knows a ward's name.
  PANE: 'misoRegionTiles',
  URL: 'tiles/outdoors/{z}/{x}/{y}.png',
  // 'css' — clip-path: path() straight on the pane. 'svg' — the same outline
  // as an SVG <clipPath> the pane points at. Measured against each other on
  // WebKit at DPR 3 before this constant was set; see region.md.
  CLIP: 'css',
  ID: 'misoRegionClip',

  layer: null,
  built: null,      // the region code the layer's bounds were made for
  data: null,       // our own copy of the file, only if /boundaries has none
  loading: false,
  key: '',          // zoom + pixel origin + region, so a pan does no work
  svg: null,
  path: null,

  // ---- what the page says ------------------------------------------------
  // the marker /region's render writes on every paint. Not a bridged state
  // key: one of those is republished at /payload's older link and would be a
  // turn behind a write made from this node (misses.md, "navigation from the
  // wrong side").

  chosen() {
    const el = document.getElementById('misoRegion');
    return el ? (el.getAttribute('data-code') || '') : '';
  },

  // /boundaries holds the parsed file once the map has painted. The region
  // page can be reloaded straight into with no map behind it, so this node
  // can fetch the file itself — into its OWN slot, never into a sibling's.
  file() {
    if (typeof feature_Boundaries !== 'undefined' && feature_Boundaries.data) {
      return feature_Boundaries.data;
    }
    return this.data;
  },

  load() {
    if (this.loading || this.data) return;
    if (typeof feature_Boundaries === 'undefined') return;
    this.loading = true;
    fetch(feature_Boundaries.FILE)
      .then((r) => (r.ok ? r.json() : null))
      .then((d) => {
        this.loading = false;
        if (!d || !d.features || !d.features.length) return;
        this.data = d;
        this.fill();
        this.safe();      // the file has landed: build the ground now, not at
                          // whatever repaint happens to come next
      })
      .catch(() => { this.loading = false; });
  },

  // the chosen feature, or the constituency: an empty code means the whole
  // patch, and so does a code the file has never heard of — a ward that was
  // renamed out of the data must not leave the map with no ground at all.
  featureFor(code) {
    const d = this.file();
    if (!d || !d.features) return null;
    let whole = null;
    let hit = null;
    for (const f of d.features) {
      const p = f.properties || {};
      if (p.kind === 'constituency' && !whole) whole = f;
      if (code && p.code === code) hit = f;
    }
    return hit || whole;
  },

  // ---- the second ground -------------------------------------------------

  ensure() {
    if (typeof L === 'undefined') return;
    if (typeof feature_Map === 'undefined' || !feature_Map.map) return;
    const map = feature_Map.map;
    if (!map.getPane(this.PANE)) {
      // Leaflet: tiles at 200, paths at 400, markers at 600. The second
      // ground goes just above the first and under the boundary lines, so the
      // line that marks the region is drawn ON the region, never under it.
      const pane = map.createPane(this.PANE);
      pane.style.zIndex = 250;
      pane.style.pointerEvents = 'none';
      pane.style.display = 'none';        // nothing is shown before it is cut
      // the map's own beats, taken once with the pane, and calling ensure()
      // rather than cut(): when the pane is made the geojson may still be on
      // its way, so a gesture has to be able to BUILD the second ground and
      // not only re-cut one that already exists. Without this the ground waits
      // for the page's next repaint, which on a quiet map may not come — the
      // rig caught exactly that: a map opened, panned and zoomed with no
      // outdoors squares on it at all until an unrelated tap repainted.
      map.on('zoomend', () => this.safe());
      map.on('moveend', () => this.safe());
      map.on('viewreset', () => this.safe());
    }
    const f = this.featureFor(this.chosen());
    if (!f) { this.off(); this.load(); return; }
    // no box means no rings means nothing to cut, and a tile layer made with
    // a null `bounds` is an UNBOUNDED one: Leaflet's own check reads a falsy
    // bounds as "every tile is valid", so a malformed geometry would ask
    // Stadia for a screenful of the world before cut() could take the layer
    // away again. The box is therefore found before the layer is made.
    const box = this.boundsOf(f);
    if (!box) { this.off(); return; }
    const code = (f.properties || {}).code || '';
    if (!this.layer) {
      // the layer is made only once a region is known, and it is BOUNDED to
      // that region: Leaflet asks for no square outside the box, so a mask the
      // size of one ward costs one ward's worth of a metered tile budget
      // rather than a screenful.
      let url = this.URL;
      if (typeof feature_FreshTiles !== 'undefined' && feature_FreshTiles.TAG) {
        url += '?' + feature_FreshTiles.TAG;
      }
      this.layer = L.tileLayer(url, {
        pane: this.PANE,
        maxZoom: 19,
        bounds: box,
        keepBuffer: 1,
        updateWhenIdle: true,
      }).addTo(map);
      this.built = code;
    } else if (code !== this.built) {
      // a different region: the box moves with it, and redraw() drops the
      // squares of the old one and asks _isValidTile about the new
      this.layer.options.bounds = box;
      this.built = code;
      this.layer.redraw();
    }
    this.cut(f);
  },

  safe() {
    try {
      this.ensure();
    } catch (e) {
    }
  },

  boundsOf(f) {
    let lo = null;
    let hi = null;
    this.rings(f).forEach((ring) => {
      for (const c of ring) {
        if (!lo) { lo = [c[1], c[0]]; hi = [c[1], c[0]]; continue; }
        if (c[1] < lo[0]) lo[0] = c[1];
        if (c[0] < lo[1]) lo[1] = c[0];
        if (c[1] > hi[0]) hi[0] = c[1];
        if (c[0] > hi[1]) hi[1] = c[0];
      }
    });
    if (!lo) return null;
    return L.latLngBounds(lo, hi);
  },

  // a Polygon's rings, or every ring of every part of a MultiPolygon — the
  // file carries both shapes, and a ward with a detached part or a hole in it
  // is an ordinary ward, not a special case. Ring 0 of each part is its
  // outside and the rest are its holes; the even-odd rule below cuts them out
  // without this code having to know which is which.
  rings(f) {
    const g = (f || {}).geometry || {};
    if (g.type === 'Polygon') return g.coordinates || [];
    if (g.type === 'MultiPolygon') {
      let out = [];
      for (const part of g.coordinates || []) out = out.concat(part);
      return out;
    }
    return [];
  },

  // ---- the mask ----------------------------------------------------------
  // The outline is written in the PANE's own coordinate space, which is
  // exactly what latLngToLayerPoint returns. That space is why this works at
  // all: Leaflet carries a pan as a translate on the map pane and a zoom
  // animation as a translate plus a scale, and a clip-path applies before an
  // ancestor's transform — so the cut travels and scales with the very tiles
  // it is cutting, and only has to be rewritten when the pixel origin moves.
  // Hence the key: a drag recomputes nothing.

  cut(f) {
    const map = feature_Map.map;
    const pane = map.getPane(this.PANE);
    if (!pane) return;
    f = f || this.featureFor(this.chosen());
    if (!f) { this.off(); return; }
    const o = map.getPixelOrigin();
    const key = map.getZoom() + '/' + o.x + '/' + o.y + '/'
      + ((f.properties || {}).code || '');
    if (key === this.key) return;
    const d = this.outline(f);
    if (!d) { this.off(); return; }
    this.key = key;
    if (this.CLIP === 'svg') {
      this.svgCut(pane, d);
    } else {
      pane.style.clipPath = 'path(evenodd, "' + d + '")';
      pane.style.webkitClipPath = 'path(evenodd, "' + d + '")';
    }
    pane.style.display = '';
  },

  outline(f) {
    const map = feature_Map.map;
    let d = '';
    for (const ring of this.rings(f)) {
      if (!ring || ring.length < 3) continue;
      for (let i = 0; i < ring.length; i++) {
        const p = map.latLngToLayerPoint(L.latLng(ring[i][1], ring[i][0]));
        d += (i ? 'L' : 'M') + p.x.toFixed(1) + ' ' + p.y.toFixed(1);
      }
      d += 'Z';
    }
    return d;
  },

  // the same outline as a real SVG clipPath in user space. Measured against
  // the CSS road on WebKit at DPR 3: byte-identical output, so the choice was
  // made on cost and the CSS one ships. Kept because if WebKit ever changes
  // its mind about one of them, the other is the CLIP constant away.
  svgCut(pane, d) {
    if (!this.svg) {
      const NS = 'http://www.w3.org/2000/svg';
      const svg = document.createElementNS(NS, 'svg');
      svg.setAttribute('width', '0');
      svg.setAttribute('height', '0');
      svg.style.position = 'absolute';
      const cp = document.createElementNS(NS, 'clipPath');
      cp.setAttribute('id', this.ID);
      cp.setAttribute('clipPathUnits', 'userSpaceOnUse');
      const p = document.createElementNS(NS, 'path');
      p.setAttribute('clip-rule', 'evenodd');
      cp.appendChild(p);
      svg.appendChild(cp);
      document.body.appendChild(svg);
      this.svg = svg;
      this.path = p;
    }
    this.path.setAttribute('d', d);
    pane.style.clipPath = 'url(#' + this.ID + ')';
    pane.style.webkitClipPath = 'url(#' + this.ID + ')';
  },

  // no region, no file, no outline: the second ground comes off entirely and
  // the map is the map it was. The layer is REMOVED rather than hidden — a
  // hidden Leaflet layer still asks for its squares, and these are metered.
  off() {
    if (typeof feature_Map === 'undefined' || !feature_Map.map) return;
    const map = feature_Map.map;
    const pane = map.getPane(this.PANE);
    if (pane) { pane.style.display = 'none'; }
    if (this.layer && map.hasLayer(this.layer)) map.removeLayer(this.layer);
    this.layer = null;
    this.built = null;
    this.key = '';
  },

  // ---- the pills ---------------------------------------------------------
  // The container is the Rust half's; the rows are the file's. `render` is
  // compiled to wasm as well as to the server and cannot read a file at all,
  // which is the honest reason the names are drawn here — and the same reason
  // /boundaries gives for the file being the seam.

  fill() {
    const box = document.getElementById('regionPills');
    if (!box) return;
    const d = this.file();
    if (!d || !d.features) { this.load(); return; }
    const now = this.featureFor(this.chosen());
    const at = ((now || {}).properties || {}).code || '';
    const rows = [];
    for (const f of d.features) {
      const p = f.properties || {};
      if (p.kind === 'constituency') {
        rows.push({ pick: '', code: p.code || '', name: p.name || 'the whole patch',
                    whole: true });
      }
    }
    const wards = [];
    for (const f of d.features) {
      const p = f.properties || {};
      if (p.kind === 'ward' && p.code) {
        wards.push({ pick: p.code, code: p.code, name: p.name || p.code });
      }
    }
    wards.sort((a, b) => a.name.localeCompare(b.name));
    let html = '';
    for (const r of rows.concat(wards)) {
      html += '<div class="region-pill' + (r.whole ? ' whole' : '')
        + (r.code === at ? ' sel' : '') + '" data-ev="region_pick:'
        + this.esc(r.pick) + '">' + this.esc(r.name) + '</div>';
    }
    if (html === box.innerHTML) return;   // a repaint that changes nothing
    box.innerHTML = html;
  },

  esc(s) {
    return String(s).replace(/&/g, '&amp;').replace(/</g, '&lt;')
      .replace(/>/g, '&gt;').replace(/"/g, '&quot;');
  },

  paint() {
    this.fill();
    this.ensure();
  },
};

{
  // no pencil on the region page. /editing/toolbar draws its edit control for
  // whatever feature_Editing.page() answers, and the region page is a
  // .card-page too — it borrows the card's ground, its insets and its scroll
  // rather than restating them. Answer nothing for it, and the toolbar's own
  // apply() takes the button away on the next paint; card pages are untouched.
  // This is /doors' idiom for the invite page, one node along (#p70): the node
  // that owns the page is the node that opts it out.
  if (typeof feature_Editing !== 'undefined' && feature_Editing.page) {
    const fm_rPageWas = feature_Editing.page.bind(feature_Editing);
    feature_Editing.page = function () {
      const p = fm_rPageWas();
      return (p && p.classList.contains('region-page')) ? null : p;
    };
  }
}

{
  // the long-press word, put into /tool-words' own table from here rather
  // than into its file: a tool's line belongs with the tool, and this way it
  // arrives and leaves with the node. `region` is keyed in TOOLS because the
  // button's event is `tool_region` — it opens a level, it does not act.
  if (typeof feature_ToolWords !== 'undefined' && feature_ToolWords.TOOLS) {
    feature_ToolWords.TOOLS.region = {
      name: 'region',
      intro: 'The patch you are working: the whole constituency, or one ward. It is drawn in outdoor colours so you can see where you are.',
    };
  }
}

{
  // /loop's paint seam by property replacement at load — /map's own idiom, and
  // taken AFTER /map so the map is mounted and synced by the time this runs.
  // With /loop or /map absent there is nothing to draw on and nothing happens.
  if (typeof feature_Loop !== 'undefined') {
    const fm_rPaint = feature_Loop.paint;
    feature_Loop.paint = function (html) {
      fm_rPaint.call(this, html);
      try {
        feature_Region.paint();
      } catch (e) {
      }
    };
  }
}

File: features/miso/loop/cards/browse/map/basemap/boundaries/region/region.rs
struct feature_Region;
impl feature_Region {
    // ---- the second ground -------------------------------------------------
    // `tiles/outdoors/{z}/{x}/{y}.png` — Stadia's Outdoors style, proxied and
    // cached exactly as /tiles proxies and caches the everyday ground, in a
    // cache directory of its own so the two grounds can never answer for each
    // other. This link sits OUTSIDE /tiles' `tiles/` handler because /tiles'
    // parser takes exactly three path segments and would 404 a fourth; being
    // newer, this node's route runs first and claims its own prefix.
    //
    // Nothing here calls a /tiles helper. `tiles.rs` reads MISO_CONTEXT_DIR
    // itself rather than borrowing /remember's `context_dir()`, for the reason
    // it states — a route should not depend on another feature being ticked —
    // and this node keeps the same discipline one step further out: it is a
    // child of /boundaries in the browse tree, and a compile-time call into
    // features/miso/serve/tiles would tie the two ticks together.

    fn route(r: request) -> response {
        if r.path.starts_with("tiles/outdoors/") {
            return region_serve(r.path.clone());
        }
        existing.route(r)
    }

    // ---- where the outdoors squares live -----------------------------------

    fn region_dir() -> String {
        let home = std::env::var("HOME").unwrap_or_default();
        let base = match std::env::var("MISO_CONTEXT_DIR") {
            Ok(d) => {
                if d.is_empty() {
                    format!("{}/.miso-context", home)
                } else {
                    d
                }
            }
            Err(_) => format!("{}/.miso-context", home),
        };
        format!("{}/tiles-outdoors", base)
    }

    // ---- the source, and the key that is never in the repo -----------------
    // Three answers, in order, and the reason for three is that the key lives
    // in exactly one place on this machine and it is not a file of ours:
    //
    //   1. MISO_OUTDOORS_URL — a whole template, for a source nobody here has
    //      heard of. The same door MISO_TILE_URL is for the everyday ground.
    //   2. STADIA_KEY — the key on its own. This is the shape the reference
    //      plist (tools/com.noob.miso.plist) documents.
    //   3. the everyday ground's own url. On the live box the key is not a
    //      variable of its own at all: it rides inside MISO_TILE_URL as
    //      `?api_key=…`. So when that url names Stadia, the style segment is
    //      swapped for `outdoors` and the query — the key — is carried over
    //      untouched. This is what makes the node work on the live server and
    //      on any rig without a new secret being set anywhere.
    //
    // With none of the three, the answer is the empty string: no url, no
    // fetch, a 404 per square, and the map draws the everyday ground inside
    // the region as well as outside it. /stand-in's rule — a missing square is
    // never an error on the page.

    fn region_source() -> String {
        match std::env::var("MISO_OUTDOORS_URL") {
            Ok(u) => {
                if !u.is_empty() {
                    return u;
                }
            }
            Err(_) => {}
        }
        match std::env::var("STADIA_KEY") {
            Ok(k) => {
                if !k.is_empty() {
                    return format!("{}?api_key={}", region_stadia_url(), k);
                }
            }
            Err(_) => {}
        }
        region_from_ground()
    }

    fn region_stadia_url() -> String {
        "https://tiles.stadiamaps.com/tiles/outdoors/{z}/{x}/{y}.png".to_string()
    }

    // the everyday ground's url with its style segment swapped. Only ever a
    // Stadia url is rewritten: any other source is left alone and answers the
    // empty string, because a style name is not a thing every tile server has.
    fn region_from_ground() -> String {
        let ground = std::env::var("MISO_TILE_URL").unwrap_or_default();
        let mark = "stadiamaps.com/tiles/";
        let at = match ground.find(mark) {
            Some(i) => i + mark.len(),
            None => return String::new(),
        };
        let rest = ground[at..].to_string();
        let end = match rest.find('/') {
            Some(i) => i,
            None => return String::new(),
        };
        format!("{}outdoors{}", &ground[..at], &rest[end..])
    }

    fn region_agent() -> String {
        match std::env::var("MISO_TILE_AGENT") {
            Ok(a) => {
                if a.is_empty() {
                    region_default_agent()
                } else {
                    a
                }
            }
            Err(_) => region_default_agent(),
        }
    }

    fn region_default_agent() -> String {
        "miso/1.0 (https://miso.xn--nb-lkaa.org; ash.nehru@gmail.com)".to_string()
    }

    // ---- the coordinates ---------------------------------------------------
    // "tiles/outdoors/{z}/{x}/{y}.png" -> [z, x, y], and the empty vector for
    // anything else. Digits only, every character checked, so no path this
    // accepts can name a file outside the cache directory — /tiles' whole
    // security argument, restated here rather than borrowed, because a parser
    // a route depends on should live beside the route. A Vec and not a tuple:
    // the chain parser cannot read a comma-bearing return type.

    fn region_coords(path: String) -> Vec<u32> {
        let rest = match path.strip_prefix("tiles/outdoors/") {
            Some(s) => s.to_string(),
            None => return Vec::new(),
        };
        let rest = match rest.strip_suffix(".png") {
            Some(s) => s.to_string(),
            None => return Vec::new(),
        };
        let parts: Vec<String> = rest.split('/').map(|p| p.to_string()).collect();
        if parts.len() != 3 {
            return Vec::new();
        }
        let mut out: Vec<u32> = Vec::new();
        for p in parts.iter() {
            if p.is_empty() || p.len() > 8 {
                return Vec::new();
            }
            if !p.chars().all(|c| c.is_ascii_digit()) {
                return Vec::new();
            }
            match p.parse::<u32>() {
                Ok(n) => out.push(n),
                Err(_) => return Vec::new(),
            }
        }
        if out[0] > 19 {
            return Vec::new();
        }
        let side = 1u32 << out[0];
        if out[1] >= side || out[2] >= side {
            return Vec::new();
        }
        out
    }

    // ---- disk first, upstream once -----------------------------------------

    fn region_serve(path: String) -> response {
        let c = region_coords(path);
        if c.len() != 3 {
            return text_response(404, "not found");
        }
        let dir = format!("{}/{}/{}", region_dir(), c[0], c[1]);
        let file = format!("{}/{}.png", dir, c[2]);
        match std::fs::read(file.clone()) {
            Ok(bytes) => {
                if !bytes.is_empty() {
                    println!("miso: outdoors {}/{}/{} disk", c[0], c[1], c[2]);
                    return region_response(bytes);
                }
            }
            Err(_) => {}
        }
        let bytes = region_fetch(c[0], c[1], c[2]);
        if !region_is_png(&bytes) {
            println!("miso: outdoors {}/{}/{} missing", c[0], c[1], c[2]);
            return text_response(404, "not found");
        }
        let _ = std::fs::create_dir_all(dir);
        let _ = std::fs::write(file, bytes.clone());
        println!("miso: outdoors {}/{}/{} fetched {} bytes", c[0], c[1], c[2],
                 bytes.len());
        region_response(bytes)
    }

    // TLS is curl's problem — /vonage's precedent, and the reason this route
    // adds no crate. No source configured is not an error either: it is a
    // fetch that is never made.
    fn region_fetch(z: u32, x: u32, y: u32) -> Vec<u8> {
        let src = region_source();
        if src.is_empty() {
            return Vec::new();
        }
        let url = src
            .replace("{z}", &z.to_string())
            .replace("{x}", &x.to_string())
            .replace("{y}", &y.to_string());
        let out = std::process::Command::new("curl")
            .arg("-s")
            .arg("-f")
            .arg("-L")
            .arg("--connect-timeout").arg("4")
            .arg("--max-time").arg("10")
            .arg("-A").arg(region_agent())
            .arg(url)
            .output();
        match out {
            Ok(o) => {
                if o.status.success() {
                    o.stdout
                } else {
                    Vec::new()
                }
            }
            Err(_) => Vec::new(),
        }
    }

    // the eight magic bytes: a captive portal's login page, or Stadia's own
    // "over your quota" json, must never be cached under a .png name.
    fn region_is_png(bytes: &Vec<u8>) -> bool {
        bytes.len() > 8 && bytes[0] == 137 && bytes[1] == 80 && bytes[2] == 78
            && bytes[3] == 71 && bytes[4] == 13 && bytes[5] == 10
            && bytes[6] == 26 && bytes[7] == 10
    }

    fn region_response(bytes: Vec<u8>) -> response {
        response { status: 200, ctype: "image/png".to_string(), body: bytes,
                   set_cookie: String::new(),
                   cache: "public, max-age=604800".to_string() }
    }

    // ---- the chosen region -------------------------------------------------
    // an ONS code, or the empty string for "the whole constituency". Empty is
    // the default because nothing about Sevenoaks belongs in this tree —
    // /boundaries' rule that the FILE is the seam. The page resolves an empty
    // code to whichever feature in the file says it is the constituency.
    //
    // Read through with_context rather than off the bridged loop state: this
    // node is newer than /payload, so a state key it published is one turn
    // behind a write made here (misses.md, "navigation from the wrong side").

    fn region_read() -> String {
        with_context(|c| c.region_region_get())
    }

    // the closure runs twice — the live world, then the turn's frozen view —
    // so it clones rather than moves. /tools' idiom.
    fn region_write(code: String) {
        edit_context(|c| {
            let _ = c.edit_op(
                "miso/loop/cards/browse/map/basemap/boundaries/region",
                "region", serde_json::json!(code.clone()));
        });
    }

    // ---- the control -------------------------------------------------------
    // A sub-tool of the posts tool, in the control row and never a page button
    // (/tools' tree-of-tools rule). It shows while the posts tool is open on
    // the MAP view with no card page up — /recentre's gate, for /recentre's
    // reason: a control that changes what the ground looks like has nothing to
    // say over a grid of tiles.
    //
    // While the region page is itself open the same button is drawn lit, as
    // the current level's own icon; the row's ‹ is /back's, and /one-level
    // carries the climb back to posts without this node saying anything —
    // `region` is deliberately kept out of `tools_list`, which is the whole
    // definition of a nested tool.

    fn tool_controls(state: String) -> String {
        let row = existing.tool_controls(state.clone());
        let open = open_tool_read();
        let mut mine = String::new();
        if open == "posts" && browse_open_read().is_empty()
            && browse_view_read() == "map" {
            mine.push_str(&region_button(false));
        }
        if open == "region" {
            mine.push_str(&region_button(true));
        }
        region_before_undo(row, mine)
    }

    // a colour of its own, /ember's deterministic pick for the name "region",
    // and NOT the posts pink. /glyphs' rule is that two controls side by side
    // must read as two things; the row already holds the posts bubble and
    // /posts' pink plus, and a third pink beside them was a wall of one colour
    // on the rig. /recentre — the other map act that lives in whatever row is
    // open — made the same call with its own name.
    fn region_button(lit: bool) -> String {
        let colour = tool_colour("region".to_string());
        let tint = if colour.is_empty() {
            String::new()
        } else {
            format!(" tinted\" style=\"--tool-colour:{}", colour)
        };
        let sel = if lit { " sel" } else { "" };
        format!("<div class=\"tool-button ctrl{}{}\" data-ev=\"tool_region\" title=\"region\">{}</div>",
                sel, tint, region_svg())
    }

    // undo is last in every row and a newer node's links land after undo's, so
    // keeping the invariant is the newcomer's job (/glyphs). Written out here
    // rather than borrowed, so this node stands whichever siblings are ticked.
    fn region_before_undo(row: String, add: String) -> String {
        if add.is_empty() {
            return row;
        }
        match row.find("data-ev=\"ctx_undo\"") {
            Some(at) => match row[..at].rfind("<div") {
                Some(start) => format!("{}{}{}", &row[..start], add, &row[start..]),
                None => format!("{}{}", row, add),
            },
            None => format!("{}{}", row, add),
        }
    }

    // ---- the events --------------------------------------------------------

    fn update(state: String, event: String) -> String {
        let state = existing.update(state, event.clone());
        let e: serde_json::Value = serde_json::from_str(&event)
            .unwrap_or(serde_json::Value::Null);
        if e["type"].as_str().unwrap_or("") != "click" {
            return state;
        }
        let ev = e["ev"].as_str().unwrap_or("").to_string();
        if ev == "region_pick" {
            region_write(String::new());
            return state;
        }
        if let Some(code) = ev.strip_prefix("region_pick:") {
            region_write(code.to_string());
        }
        state
    }

    // ---- the surface -------------------------------------------------------
    // Two pieces of html, and neither of them names a ward.
    //
    // The pills are an EMPTY container. The names and the codes live in
    // /boundaries' geojson, which is a file on the page — and `render` is
    // compiled to wasm as well as to the server, so it cannot read a file at
    // all. The page half fills the container from the parsed collection
    // /boundaries already holds. That division is not a workaround: the file
    // is the seam (/boundaries), and this keeps Sevenoaks out of the Rust as
    // firmly as it is out of the JavaScript.
    //
    // `#misoRegion` is how the map half learns the choice, and it is a marker
    // in the html rather than a bridged state key for the reason /browse
    // states: a key republished at /payload's older link would paint one stale
    // frame after a write made here. This element is written by the render
    // that follows this node's own update, so it is never a turn behind.

    fn render(state: String) -> String {
        let base = existing.render(state.clone());
        let code = region_read();
        let mark = format!("<span id=\"misoRegion\" data-code=\"{}\" hidden></span>",
                           region_esc(code));
        if open_tool_read() != "region" {
            return format!("{}{}", base, mark);
        }
        format!("{}{}<div class=\"card-page region-page\"><div id=\"regionPills\" class=\"region-pills\"></div></div>",
                base, mark)
    }

    fn region_esc(s: String) -> String {
        s.replace('&', "&amp;").replace('<', "&lt;").replace('>', "&gt;")
            .replace('"', "&quot;")
    }

    // ---- the glyph ---------------------------------------------------------
    // drawn in currentColor, never an emoji (/glyphs): a patch of ground with
    // one part of it filled in — which is the whole of what this tool does.
    //
    // Not a folded map. The first cut was one, and on the rig it read as the
    // picker's own map glyph three fingers away — the same shape, twice, for
    // two different things. This is an irregular patch with one internal
    // division and one side solid: an area, and the part of it that is yours.

    fn region_svg() -> String {
        String::from(concat!(
            "<svg class=\"icon-svg\" viewBox=\"0 0 24 24\" aria-hidden=\"true\">",
            "<path d=\"M4 6.8 11 3.4 12.2 20.6 4.6 17.6z\" ",
            "fill=\"currentColor\" fill-opacity=\"0.45\"/>",
            "<path d=\"M4 6.8 11 3.4 20 6.2 19.4 17.4 12.2 20.6 4.6 17.6z\" ",
            "fill=\"none\" stroke=\"currentColor\" stroke-width=\"2.2\" ",
            "stroke-linecap=\"round\" stroke-linejoin=\"round\"/>",
            "<path d=\"M11 3.4 12.2 20.6\" fill=\"none\" ",
            "stroke=\"currentColor\" stroke-width=\"2.2\" ",
            "stroke-linecap=\"round\" stroke-linejoin=\"round\"/>",
            "</svg>"))
    }
}
